### ISSUE TITLE
DOC: general discussion of our docstring sharing system

### ISSUE THREAD

**Issue by jorisvandenbossche** [pandas-dev/pandas#19932]
Currently in many places we write docstrings as string values inside a `_shared_docs` dictionary, and then add this docstring to the method with a decorator like `@Appender(_shared_docs['where'])`, or possibly with substitution of variables in the template docstring with `@Appender(_shared_docs['where'] % _shared_doc_kwargs)`. 
Apart from that, we have some other custom ways to generate them, typically for a specific set of methods (eg all statistical methods like mean, sum, median, ..).

This system is important to not duplicate content too much, but also has some drawbacks:

- I think it is a rather complex system for contributors to understand. It can be hard to know where to find the method's docstring, or how to edit it appropriately, and how it will look like eventually.
- It gives rise to some errors (which can of course be solved). For example:
  - left-overs from shared_kwargs of parent docstrings that have not been substituted correctly. Eg NDFrame should never occur in the docs, but does quite a lot (eg [DataFrame.where docstring](http://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.DataFrame.where.html))
  - left-overs of the template where we forgot to substitute at all (eg [DataFrame.rename docstring](http://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.DataFrame.rename.html))
- We have examples that are not specific for the function (eg [DataFrame.mask docstring](http://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.DataFrame.mask.html) with DataFrame.where examples) or for the object (eg [Index.searchsorted docstring](http://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.Index.searchsorted.html) with Series examples). 
 And in general in can be hard/complicated to add good examples, as those are typically specific to one method, so you need to write them in separate strings and then add to the appropriate docstring.

So therefore opening this issue, to gather ideas on whether we can improve this workflow, without sacrificing the deduplication and flexibility of the current system?

**Comment by jorisvandenbossche** [comment]
Some first thoughts:

* Better document the workflow for contributors (this is actually issue https://github.com/pandas-dev/pandas/issues/16446)
* Remove `% _shared_doc_kwargs` when there is nothing to substitute, and only put values inside `_shared_doc_kwargs` if they are used multiple times: if it is only for one method, write the string there where it is substituted.
* I was wondering, would `@Appender(NDFrame.sort_values)` be more clear for contributors compared to `@Appender(generic._shared_docs['sort_values'])` ? That would also mean that in `generic.py` the docstring can be written as a normal docstring (after the `def` line instead of before)
* We use both `@Substitution(...)` and `% _shared_doc_kwargs` (within an `Appender` decorator). Might be good to harmonize this.

**Comment by jorisvandenbossche** [comment]
A typical occurrence of a problem is a function that *only* lives in generic.py, and has a template with substituted values. But since it only lives in generic.py, it only gets substituted with 'NDFrame' et al. 
Examples of this are `pipe` (see eg https://github.com/pandas-dev/pandas/blob/master/pandas/core/generic.py#L3929), `pct_change`, `mask`, `where`, `set_axis`.

One possibility is to actually add them to `frame.py` and `series.py`, just to call the super method but to fill in the template. Like eg is done for isna and related methods:

https://github.com/pandas-dev/pandas/blob/52559f535788b3c59c6c78d2cf987b03303b16f6/pandas/core/frame.py#L3482-L3484

Other possibility is to update the docstring to just be generic (replace 'klass' with Series/DataFrame). Although then we would also need to add examples of both.

Last possibility would also be to do this template filling without overriding the actual method, but only filling the docstring with eg

```
DataFrame.isna.__doc__ = DataFrame.isna.__doc__ % _shared_doc_kwargs
```

(or go with some fancy metaclassing to do this automatically ...)

cc @jreback @TomAugspurger

**Comment by jorisvandenbossche** [comment]
Another aspect to consider: to what extent do we necessarily want separate docstrings for Series/DataFrame? 
For certain methods it might be fine to just have a common docstring (without substituting things), with examples that show both Series and DataFrame?

Anyway, some feedback on all the above thoughts is welcome :-)

**Comment by TomAugspurger** [comment]
Personally, I'm fine with using the same docstring for Series / DataFrame
in some cases, as long as the signature is identical. I would need to go
through some individual examples though.

**Comment by jorisvandenbossche** [comment]
Any opinions on this:

> I was wondering, would `@Appender(NDFrame.sort_values)` be more clear for contributors compared to `@Appender(generic._shared_docs['sort_values'])` ? That would also mean that in generic.py the docstring can be written as a normal docstring (after the def line instead of before)

**Comment by TomAugspurger** [comment]
Yes, I support that. I started the shared doc guide this morning, and noticed what you meant by that.

**Comment by max-sixty** [comment]
FYI at https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.set_axis.html there's:

> Whether to return a new %(klass)s instance.

**Comment by simonjayhawkins** [comment]
Given the work happening in #62437 and the PRs linked there, this older issue appears to be superseded. The original motivations and trade-offs discussed here have now been revisited in the newer proposal, so it makes sense to close this one to avoid duplication.
